File: leetcode/py/p22-generate-parentheses.py
# LeetCode 22. Generate parentheses.
from typing import List


def generate_parentheses_rec(l: List[str], s: List[str], left: int, right: int):
    """Backtracking recursive call to generate list of valid parentheses.

    :param l: Builds the list of strings with valid parentheses.
    :param s: The current stack of () parens.
    :param left: Number of left parens in use.
    :param right: Number of right parens in use.
    :return: None
    """
    if left > 0:
        s.append("(")
        generate_parentheses_rec(l, s, left - 1, right)
        s.pop()
    elif right == 0:
        l.append("".join(s))
        return
    if right > left:
        s.append(")")
        generate_parentheses_rec(l, s, left, right - 1)
        s.pop()


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        assert n >= 1 and n <= 8
        l = []
        generate_parentheses_rec(l, [], n, n)
        return l


if __name__ == "__main__":
    tests = (
        (1, ["()"]),
        (2, ["()()", "(())"]),
        (3, ["((()))", "(()())", "(())()", "()(())", "()()()"]),
        (
            4,
            [
                "(((())))",
                "((()()))",
                "((())())",
                "((()))()",
                "(()(()))",
                "(()()())",
                "(()())()",
                "(())(())",
                "(())()()",
                "()((()))",
                "()(()())",
                "()(())()",
                "()()(())",
                "()()()()",
            ],
        ),
    )
    solution = Solution()
    for n, exp in tests:
        exp = set(exp)
        act = set(solution.generateParenthesis(n))
        if exp != act:
            print(f"Fail. n: {n}, exp: {exp}, act: {act}")

File: leetcode/cpp/p22-generate-parentheses.cpp
// LeetCode 22. Generate parentheses.
#include "leetcode.hpp"

#include <algorithm>
#include <iostream>
#include <vector>

bool has_valid_parens(const std::string& s)
{
    int i = 0;
    for (const char c : s) {
        if (c == '(') {
            ++i;
        } else {
            --i;
            if (i < 0) {
                return false;
            }
        }
    }
    return true;
}

class Solution {
public:
    std::vector<std::string> generateParenthesis(int n)
    {
        std::vector<std::string> v;
        if (n <= 0) {
            return v;
        }
        std::string s(static_cast<size_t>(n * 2), ')');
        std::fill_n(s.begin(), static_cast<size_t>(n), '(');
        if (n == 1) {
            v.push_back(s);
            return v;
        }
        do {
            if (has_valid_parens(s)) {
                v.push_back(s);
            }
        } while (std::next_permutation(s.begin() + 1, s.end() - 1));
        return v;
    }
};

int main(int argc, char* argv[])
{
    struct TestCase {
        int n;
        std::vector<std::string> exp;
    };
    const TestCase testCases[] = {
        { 0, {} },
        { 1, { "()" } },
        { 2, { "(())", "()()" } },
        { 3, { "((()))", "(()())", "(())()", "()(())", "()()()" } },
        { 4, { "(((())))", "((()()))", "((())())", "((()))()", "(()(()))", "(()()())", "(()())()", "(())(())", "(())()()", "()((()))", "()(()())", "()(())()", "()()(())", "()()()()" } },
    };

    Solution solution;
    for (const auto& tc : testCases) {
        const auto ans = solution.generateParenthesis(tc.n);
        if (tc.exp != ans) {
            std::cout << "FAIL. " << __FUNCTION__ << "(n: " << tc.n << ")"
                      << ", exp: " << leetcode::to_string(tc.exp)
                      << ", ans: " << leetcode::to_string(ans)
                      << "\n";
        }
    }
    return 0;
}
